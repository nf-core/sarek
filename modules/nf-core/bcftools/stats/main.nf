process BCFTOOLS_STATS {
    tag "$meta.id"
    label 'process_single'

    conda "${moduleDir}/environment.yml"
    container "${ workflow.containerEngine == 'singularity' && !task.ext.singularity_pull_docker_container ?
        'https://depot.galaxyproject.org/singularity/bcftools:1.20--h8b25389_0':
        'biocontainers/bcftools:1.20--h8b25389_0' }"

    input:
    tuple val(meta),  path(vcf), path(tbi)
    tuple val(meta2), path(regions)
    tuple val(meta3), path(targets)
    tuple val(meta4), path(samples)
    tuple val(meta5), path(exons)
    tuple val(meta6), path(fasta)

    output:
    tuple val(meta), path("*stats.txt"), emit: stats
    path  "versions.yml"               , emit: versions

    when:
    task.ext.when == null || task.ext.when

    script:
    def args = task.ext.args ?: ''
    def prefix = task.ext.prefix ?: "${meta.id}"
    def regions_file = regions ? "--regions-file ${regions}" : ""
    def targets_file = targets ? "--targets-file ${targets}" : ""
    def samples_file =  samples ? "--samples-file ${samples}" : ""
    def reference_fasta = fasta ? "--fasta-ref ${fasta}" : ""
    def exons_file = exons      ? "--exons ${exons}" : ""
    """
    bcftools stats \\
        $args \\
        $regions_file \\
        $targets_file \\
        $samples_file \\
        $reference_fasta \\
        $exons_file \\
        $vcf > ${prefix}.bcftools_stats.txt

    cat <<-END_VERSIONS > versions.yml
    "${task.process}":
        bcftools: \$(bcftools --version 2>&1 | head -n1 | sed 's/^.*bcftools //; s/ .*\$//')
    END_VERSIONS
    """

    stub:
    def prefix = task.ext.prefix ?: "${meta.id}"

    """
    touch ${prefix}.bcftools_stats.txt

    cat <<-END_VERSIONS > versions.yml
    "${task.process}":
        bcftools: \$(bcftools --version 2>&1 | head -n1 | sed 's/^.*bcftools //; s/ .*\$//')
    END_VERSIONS
    """
}
