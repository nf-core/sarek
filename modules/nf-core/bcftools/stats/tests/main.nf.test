nextflow_process {

    name "Test Process BCFTOOLS_STATS"
    script "../main.nf"
    process "BCFTOOLS_STATS"

    tag "modules"
    tag "modules_nfcore"
    tag "bcftools"
    tag "bcftools/stats"

    test("sarscov2 - vcf_gz") {

        when {
            process {
                """
                input[0] = [ [ id:'test' ], // meta map
                            file(params.modules_testdata_base_path + 'genomics/sarscov2/illumina/vcf/test.vcf.gz', checkIfExists: true),
                            []]
                input[1] = [ [], [] ]
                input[2] = [ [], [] ]
                input[3] = [ [], [] ]
                input[4] = [ [], [] ]
                input[5] = [ [], [] ]
                """
            }
        }

        then {
            assertAll(
                { assert process.success },
                { assert snapshot(process.out.versions).match("versions") },
                { assert snapshot(file(process.out.stats.get(0).get(1)).readLines()[0..5]).match() },
            )
        }

    }

    test("sarscov2 - vcf_gz - regions") {

        when {
            process {
                """
                input[0] = [ [ id:'regions_test' ], // meta map
                            file(params.modules_testdata_base_path + 'genomics/sarscov2/illumina/vcf/test.vcf.gz', checkIfExists: true),
                            file(params.modules_testdata_base_path + 'genomics/sarscov2/illumina/vcf/test.vcf.gz.tbi', checkIfExists: true)]
                input[1] = [ [id:'regions_test'],
                            file(params.modules_testdata_base_path + 'genomics/sarscov2/illumina/vcf/test3.vcf.gz', checkIfExists: true) ]
                input[2] = [ [], [] ]
                input[3] = [ [], [] ]
                input[4] = [ [], [] ]
                input[5] = [ [], [] ]
                """
            }
        }

        then {
            assertAll(
                { assert process.success },
                { assert snapshot(process.out.versions).match("regions_versions") },
                { assert snapshot(file(process.out.stats.get(0).get(1)).readLines()[0..5]).match() },
            )
        }

    }

    test("sarscov2 - vcf_gz - targets") {

        when {
            process {
                """
                input[0] = [ [ id:'targets_test' ], // meta map
                            file(params.modules_testdata_base_path + 'genomics/sarscov2/illumina/vcf/test.vcf.gz', checkIfExists: true),
                            [] ]
                input[1] = [ [], [] ]
                input[2] = [ [id:'targets_test'],
                            file(params.modules_testdata_base_path + 'genomics/sarscov2/illumina/vcf/test2.targets.tsv.gz', checkIfExists: true)
                            ]
                input[3] = [ [], [] ]
                input[4] = [ [], [] ]
                input[5] = [ [], [] ]
                """
            }
        }

        then {
            assertAll(
                { assert process.success },
                { assert snapshot(process.out.versions).match("targets_versions") },
                { assert snapshot(file(process.out.stats.get(0).get(1)).readLines()[0..5]).match() },
            )
        }

    }

    test("sarscov2 - vcf_gz - exons") {

        when {
            process {
                """
                input[0] = [ [ id:'exon_test' ], // meta map
                            file(params.modules_testdata_base_path + 'genomics/sarscov2/illumina/vcf/test.vcf.gz', checkIfExists: true),
                            [] ]
                input[1] = [ [], [] ]
                input[2] = [ [], [] ]
                input[3] = [ [], [] ]
                input[4] = [ [id: "exon_test"],
                            file(params.modules_testdata_base_path + 'delete_me/bcftools/stats/exons.tsv.gz', checkIfExists: true) ]
                input[5] = [ [], [] ]
                """
            }
        }

        then {
            assertAll(
                { assert process.success },
                { assert snapshot(process.out.versions).match("exon_versions") },
                { assert snapshot(file(process.out.stats.get(0).get(1)).readLines()[0..5]).match() },
            )
        }

    }

    test("sarscov2 - vcf_gz - reference") {

        when {
            process {
                """
                input[0] = [ [ id:'ref_test' ], // meta map
                            file(params.modules_testdata_base_path + 'genomics/sarscov2/illumina/vcf/test.vcf.gz', checkIfExists: true),
                            [] ]
                input[1] = [ [], [] ]
                input[2] = [ [], [] ]
                input[3] = [ [], [] ]
                input[4] = [ [], [] ]
                input[5] = [ [id: 'ref_test'],
                            file(params.modules_testdata_base_path + 'genomics/sarscov2/genome/genome.fasta', checkIfExists: true)
                            ]
                """
            }
        }

        then {
            assertAll(
                { assert process.success },
                { assert snapshot(process.out.versions).match("ref_versions") },
                { assert snapshot(file(process.out.stats.get(0).get(1)).readLines()[0..5]).match() },
            )
        }

    }


    test("sarscov2 - vcf_gz - stub") {

        options "-stub"

        when {
            process {
                """
                input[0] = [ [ id:'test' ], // meta map
                            file(params.modules_testdata_base_path + 'genomics/sarscov2/illumina/vcf/test.vcf.gz', checkIfExists: true),
                            []]
                input[1] = [ [], [] ]
                input[2] = [ [], [] ]
                input[3] = [ [], [] ]
                input[4] = [ [], [] ]
                input[5] = [ [], [] ]
                """
            }
        }

        then {
            assertAll(
                { assert process.success },
                { assert snapshot(process.out).match() }
            )
        }

    }

}