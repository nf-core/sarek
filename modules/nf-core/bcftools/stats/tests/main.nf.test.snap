{
    "sarscov2 - vcf_gz - reference": {
        "content": [
            [
                "# This file was produced by bcftools stats (1.20+htslib-1.20) and can be plotted using plot-vcfstats.",
                "# The command line was:\tbcftools stats  --fasta-ref genome.fasta test.vcf.gz",
                "#",
                "# Definition of sets:",
                "# ID\t[2]id\t[3]tab-separated file names",
                "ID\t0\ttest.vcf.gz"
            ]
        ],
        "meta": {
            "nf-test": "0.8.4",
            "nextflow": "23.10.1"
        },
        "timestamp": "2024-05-31T18:14:35.506777837"
    },
    "sarscov2 - vcf_gz - exons": {
        "content": [
            [
                "# This file was produced by bcftools stats (1.20+htslib-1.20) and can be plotted using plot-vcfstats.",
                "# The command line was:\tbcftools stats  --exons exons.tsv.gz test.vcf.gz",
                "#",
                "# Definition of sets:",
                "# ID\t[2]id\t[3]tab-separated file names",
                "ID\t0\ttest.vcf.gz"
            ]
        ],
        "meta": {
            "nf-test": "0.8.4",
            "nextflow": "23.10.1"
        },
        "timestamp": "2024-05-31T18:14:30.57486244"
    },
    "versions": {
        "content": [
            [
                "versions.yml:md5,17cdf9d1ad31f6b1f5935dfcc9fe7b9a"
            ]
        ],
        "meta": {
            "nf-test": "0.8.4",
            "nextflow": "23.10.1"
        },
        "timestamp": "2024-05-31T18:16:27.637515559"
    },
    "sarscov2 - vcf_gz - targets": {
        "content": [
            [
                "# This file was produced by bcftools stats (1.20+htslib-1.20) and can be plotted using plot-vcfstats.",
                "# The command line was:\tbcftools stats  --targets-file test2.targets.tsv.gz test.vcf.gz",
                "#",
                "# Definition of sets:",
                "# ID\t[2]id\t[3]tab-separated file names",
                "ID\t0\ttest.vcf.gz"
            ]
        ],
        "meta": {
            "nf-test": "0.8.4",
            "nextflow": "23.10.1"
        },
        "timestamp": "2024-05-31T18:14:25.732997442"
    },
    "regions_versions": {
        "content": [
            [
                "versions.yml:md5,17cdf9d1ad31f6b1f5935dfcc9fe7b9a"
            ]
        ],
        "meta": {
            "nf-test": "0.8.4",
            "nextflow": "23.10.1"
        },
        "timestamp": "2024-05-31T18:16:32.559884458"
    },
    "targets_versions": {
        "content": [
            [
                "versions.yml:md5,17cdf9d1ad31f6b1f5935dfcc9fe7b9a"
            ]
        ],
        "meta": {
            "nf-test": "0.8.4",
            "nextflow": "23.10.1"
        },
        "timestamp": "2024-05-31T18:16:37.512009805"
    },
    "sarscov2 - vcf_gz - stub": {
        "content": [
            {
                "0": [
                    [
                        {
                            "id": "test"
                        },
                        "test.bcftools_stats.txt:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "1": [
                    "versions.yml:md5,17cdf9d1ad31f6b1f5935dfcc9fe7b9a"
                ],
                "stats": [
                    [
                        {
                            "id": "test"
                        },
                        "test.bcftools_stats.txt:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "versions": [
                    "versions.yml:md5,17cdf9d1ad31f6b1f5935dfcc9fe7b9a"
                ]
            }
        ],
        "meta": {
            "nf-test": "0.8.4",
            "nextflow": "24.04.2"
        },
        "timestamp": "2024-06-03T11:57:09.614976125"
    },
    "exon_versions": {
        "content": [
            [
                "versions.yml:md5,17cdf9d1ad31f6b1f5935dfcc9fe7b9a"
            ]
        ],
        "meta": {
            "nf-test": "0.8.4",
            "nextflow": "23.10.1"
        },
        "timestamp": "2024-05-31T18:16:42.347397266"
    },
    "ref_versions": {
        "content": [
            [
                "versions.yml:md5,17cdf9d1ad31f6b1f5935dfcc9fe7b9a"
            ]
        ],
        "meta": {
            "nf-test": "0.8.4",
            "nextflow": "23.10.1"
        },
        "timestamp": "2024-05-31T18:16:47.26823622"
    },
    "sarscov2 - vcf_gz": {
        "content": [
            [
                "# This file was produced by bcftools stats (1.20+htslib-1.20) and can be plotted using plot-vcfstats.",
                "# The command line was:\tbcftools stats  test.vcf.gz",
                "#",
                "# Definition of sets:",
                "# ID\t[2]id\t[3]tab-separated file names",
                "ID\t0\ttest.vcf.gz"
            ]
        ],
        "meta": {
            "nf-test": "0.8.4",
            "nextflow": "23.10.1"
        },
        "timestamp": "2024-05-31T18:16:27.670416598"
    },
    "sarscov2 - vcf_gz - regions": {
        "content": [
            [
                "# This file was produced by bcftools stats (1.20+htslib-1.20) and can be plotted using plot-vcfstats.",
                "# The command line was:\tbcftools stats  --regions-file test3.vcf.gz test.vcf.gz",
                "#",
                "# Definition of sets:",
                "# ID\t[2]id\t[3]tab-separated file names",
                "ID\t0\ttest.vcf.gz"
            ]
        ],
        "meta": {
            "nf-test": "0.8.4",
            "nextflow": "23.10.1"
        },
        "timestamp": "2024-05-31T18:14:20.759094062"
    }
}